src/io/MemoryProfiler.cpp

```diff
@@ -67,72 +63,46 @@ void MemoryProfiler::doOutput(const std::string& myString) {
 
 unsigned long long MemoryProfiler::getCachedSize() {
 	unsigned long long cached_size = 0;
-	const size_t MAXLEN = 1024;
-	FILE *fp;
-	char buf[MAXLEN];
-	fp = fopen("/proc/meminfo", "r");
-	while (fgets(buf, MAXLEN, fp)) {
-		char *p1 = strstr(buf, "Cached:");
-		if (p1 != NULL) {
-			int colon = ':';
-			p1 = strchr(buf, colon) + 1;
-			cached_size = strtoull(p1, NULL, 10);
+	std::ifstream meminfo_file("/proc/meminfo");
+	std::string token;
+	while(meminfo_file >> token) {
+		if (token == "Cached:") {
+			meminfo_file >> cached_size;
 			break;
 		}
 	}
-	fclose(fp);
 	return cached_size;
 }
 
-int MemoryProfiler::parseLine(char* line) {
-	// This assumes that a digit will be found and the line ends in " Kb".
-	int i = strlen(line);
-	const char* p = line;
-	while (*p < '0' || *p > '9')
-		p++;
-	line[i - 3] = '\0';
-	i = atoi(p);
-	return i;
-}
-
 int MemoryProfiler::countHugePages() {
-	FILE* file = fopen("/proc/self/numa_maps", "r");
-	if (!file) {
-		return 0;
-	}
+	std::ifstream numa_maps_file("/proc/self/numa_maps");
+
 	int hugepagecount = 0;
-	char line[1024];
-	while (fgets(line, 1024, file) != NULL) {
+	std::string line;
+	while(std::getline(numa_maps_file, line)) {
 		//if (/huge.*dirty=(\d+)/) {
-		std::string linestring(line);
-		if (linestring.find("huge") != std::string::npos) {
-			auto pos = linestring.find("dirty");
-			if (pos != std::string::npos) {
-				linestring = linestring.substr(pos);
-				auto eqpos = linestring.find("=");
-				auto spacepos = linestring.find(" ");
-				linestring = linestring.substr(eqpos + 1, spacepos - eqpos - 1);
-				hugepagecount += std::stoi(linestring);
+		if (line.find("huge") != std::string::npos) {
+			const std::regex re("dirty=(\\d+) ");
+			std::smatch match;
+			if (std::regex_match(line, match, re)) {
+				hugepagecount += std::stoi(match[1].str());
 			}
 		}
 	}
-	fclose(file);
 	return hugepagecount;
 }
 
 int MemoryProfiler::getOwnMemory() { //Note: this value is in KB!
-	FILE* file = fopen("/proc/self/status", "r");
-	int result = -1;
-	char line[1024];
-
-	while (fgets(line, 128, file) != NULL) {
-		if (strncmp(line, "VmRSS:", 6) == 0) {
-			result = parseLine(line);
+	int vmrss = 0;
+	std::ifstream self_status_file("/proc/self/status");
+	std::string token;
+	while(self_status_file >> token) {
+		if (token == "VmRSS:") {
+			self_status_file >> vmrss;
```

Comment:
While the previous implementation was a bit too complicated (`MemoryProfiler::parseLine()`), this _might_ be a bit too simple. We could think of some additional checks (also at every other location where external input is processed/parsed). But I am also fine with your current improvement if it is kind of guaranteed that the input files are **always** structured in this way.

Reply:
The format of /proc/<PID>/status is shown in the linux documentation (https://man7.org/linux/man-pages/man5/proc_pid_status.5.html) - if it changes lots of other (system) programs are likely to break, so this should be fix.